File: CPlusPlus/QuickSort/QuickSort-T1.cpp
#include <iostream>
using namespace std;

void funcSwap(int &a, int &b)
{
	int tmp = a;
	a = b;
	b = tmp;
}

int funcPartition(int lst[], int left, int right)
{
	int i = left;
	int pivot = lst[left];
	int j = left;
	while (++i <= right)
	{
		if (lst[i] < pivot)
		{
			j++;
			if (i != j)
			{
				funcSwap(lst[i], lst[j]);
			}
		}
	}
	if (j != left)
	{
		funcSwap(lst[left], lst[j]);
	}
	return j;
}

void QuickSort(int lst[], int left, int right)
{
	if (left >= right)
		return;
	int idx = funcPartition(lst, left, right);
	QuickSort(lst, left, idx - 1);
	QuickSort(lst, idx + 1, right);
}

int main(int argc, char *argv[])
{
	cout << "QuickSort-T1 is running ..." << endl;

	int lst[] = {2, 4, 6, 8, 0, 9, 7, 5, 3, 1};
  // int lst[] = {12, 45, 9, 748, 12, 56, 3, 89, 4, 48, 2};
	int len = sizeof(lst) / sizeof(int);

	QuickSort(lst, 0, len - 1);

	for (int i = 0; i < len; i++)
	{
		cout << lst[i] << " ";
	}
}

File: Nodejs/QuickSortTest/QuickSort-T1.js
#!/usr/bin/env node

const funcSwap = (lst, i, j) => {
	const tmp = lst[i];
	lst[i] = lst[j];
	lst[j] = tmp;
};


const pickMiddleIdx = (lst, left, right) => {
	const middle = Math.floor((left + right) / 2);
	if ((lst[left] - lst[middle]) * (lst[middle] - lst[right]) >= 0) {
		return middle;
	} else if ((lst[middle] - lst[left]) * (lst[left] - lst[right]) >= 0) {
		return left;
	} else {
		return right;
	}
}


const funcPartition = (lst, left, right) => {
	const pIdx = pickMiddleIdx(lst, left, right);
	if (pIdx !== left) {
		funcSwap(lst, left, pIdx);
	}
	const pivot = lst[left];
	let j = left;
	for (let i = left + 1; i <= right; i++) {
		if (lst[i] < pivot) {
			j++;
			if (i !== j) {
				funcSwap(lst, i, j);
			}
		}
	}
	if (j !== left) {
		funcSwap(lst, left, j);
	}
	return j;
};


const quickSort = (lst, left, right) => {
	if (left >= right) {
		return;
	}

	const idx = funcPartition(lst, left, right);
	quickSort(lst, left, idx - 1);
	quickSort(lst, idx + 1, right);
};


const lst = [ 5, 4, 6, 8, 2, 9, 7, 1, 3, 0 ];

quickSort(lst, 0, lst.length - 1);

console.log(`lst = ${lst}`);

